### PR TITLE
Kernel: Add support for opting-out from compiling various drivers

### PR DESCRIPTION
**Pull request by supercomputer7**
This is needed so we can compile smaller Kernels, to ensure we can reduce boot times on bare metal, and to not include unnecessary drivers for specific platforms (for example, my ICH7 test machine doesn't have AHCI support, therefore I will create a Kernel without it).

### CODE REVIEW

**Comment by Hendiadyoin1** [comment]
It would be cool if this could harness the `configure-components` helper-system.

**Comment by supercomputer7** [comment]
> It would be cool if this could harness the `configure-components` helper-system.

I agree, but it's not in the scope of this PR.

**Comment by supercomputer7** [comment]
2 important things to fix:
1. I want to fix is where the config file sits, I prefer that the config file will be a fixed one in the Kernel folder, but CMake will actually use the one in the Build directory.
2.  I want to create a CI that enables all compile flags of the Kernel, to ensure that even if the default file disables some features (e.g. RTL8139 support, which we don't use normally), we don't regress anything.

File: Kernel/Modules.cmake [hunk 1]
```diff
@@ -0,0 +1,162 @@
+set_property(DIRECTORY APPEND PROPERTY CMAKE_CONFIGURE_DEPENDS "${CMAKE_SOURCE_DIR}/Kernel/config.ini")
+
+macro( readSettingFile KEY DEFAULT_RESULT STRING_RESULT_OUT)
```
Comment (ADKaster) [inline Kernel/Modules.cmake:3]: It's unfortunate that this reads the file from 0 for each variable we want to look up.

I think a simpler design if we want to keep the separate ini file might be something like

```cmake
file(READ "${CMAKE_SOURCE_DIR}/Kernel/config.ini" contents )
string(REGEX REPLACE ";" "\\\\;" contents "${contents}")
string(REGEX REPLACE "\n" ";" contents "${contents}")
foreach(line IN LISTS contents)
    string(REPLACE "=" ";" line_list line)
    list(GET line_list 0 var)
    list(GET line_list 1 value)
    set("${var}" "${value}" CACHE BOOL FORCE) # force the changes from the file to take effect
endforeach()
```

If you did this, then you can just check 
```
if (COMPILE_RAMDISK)
   # add ramdisk cpp and define
endif()
```

File: Kernel/Net/NetworkingManagement.cpp [hunk 2]
```diff
@@ -75,16 +85,26 @@ RefPtr<NetworkAdapter> NetworkingManagement::lookup_by_name(const StringView& na
...
 {
+#ifdef COMPILE_E1000
     if (auto candidate = E1000NetworkAdapter::try_to_initialize(address); !candidate.is_null())
         return candidate;
+#endif
+#ifdef COMPILE_E1000E
     if (auto candidate = E1000ENetworkAdapter::try_to_initialize(address); !candidate.is_null())
         return candidate;
+#endif
+#ifdef COMPILE_RTL8139
```
Comment (ADKaster) [inline Kernel/Net/NetworkingManagement.cpp:96]: Seeing the changes to the Networking and Graphics code makes me even less thrilled about the extra #ifdefs added as  a part of this modularization. We discussed a few  options in discord, such as

- keep the ifdefs
- `initalize_if(COMPILE_XXXX)` or similar helpers to hide the incomplete types of non-compiled-in drivers behind an ifdef.
- registering function pointers in the .cpp file for the associated hardware device to perform initialization when requested.
- using global constructors to make FooDeviceX initialization available to FooManagement initialization. 

I'm not sure which one is the best design. These all feel like they could use some design pattern elbow grease. Some factories, or interfaces, or something else to decouple the Managment classes from the concrete implementation classes of each hardware device.

File: Kernel/CMakeLists.txt [hunk 3]
```diff
@@ -451,6 +430,9 @@ if (NOT "${SERENITY_ARCH}" STREQUAL "aarch64")
     add_library(kernel_heap STATIC ${KERNEL_HEAP_SOURCES})
 endif()
 add_executable(Kernel ${SOURCES})
+
+include("${CMAKE_SOURCE_DIR}/Kernel/Modules.cmake")
```
Comment (ADKaster) [inline Kernel/CMakeLists.txt:434]: Probably best to not hard-code the project structure in the CMake lists. I did put some effort into making all of the `.cmake` files live in Meta/CMake though, so moving the file there and calling it `KernelModules.cmake` or some such would let us `include(KernelModules)`

```suggestion
include("${CMAKE_CURRENT_LIST_DIR}/Modules.cmake")
```

File: Kernel/Modules.cmake [hunk 4]
```diff
@@ -0,0 +1,162 @@
+if (NOT EXISTS "${CMAKE_SOURCE_DIR}/Build/${SERENITY_ARCH}/Kernel/config.ini")
+    configure_file("${CMAKE_SOURCE_DIR}/Kernel/config.ini.template" "${CMAKE_SOURCE_DIR}/Build/${SERENITY_ARCH}/Kernel/config.ini" COPYONLY)
+endif()
+set_property(DIRECTORY APPEND PROPERTY CMAKE_CONFIGURE_DEPENDS "${CMAKE_SOURCE_DIR}/Build/${SERENITY_ARCH}/Kernel/config.ini")
```
Comment (ADKaster) [inline Kernel/Modules.cmake:4]: This is a bit awkward with the hard-coded binary directory. namely, it won't work when you have a Clang build. 

```suggestion
if (NOT EXISTS "${CMAKE_CURRENT_BINARY_DIR}/config.ini")
    configure_file("${CMAKE_SOURCE_DIR}/Kernel/config.ini.template" "${CMAKE_CURRENT_BINARY_DIR}/config.ini" COPYONLY)
endif()
set_property(DIRECTORY APPEND PROPERTY CMAKE_CONFIGURE_DEPENDS "${CMAKE_CURRENT_BINARY_DIR}/config.ini")
```

File: Kernel/Modules.cmake [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+# Default values for default setup
+set(COMPILE_IDE "y")
+set(COMPILE_AHCI "y")
+set(COMPILE_RAMDISK "y")
+
+set(COMPILE_E1000 "y")
+set(COMPILE_E1000E "y")
+set(COMPILE_RTL8139 "y")
+set(COMPILE_RTL8168 "y")
+set(COMPILE_NE2000 "y")
```
Comment (ADKaster) [inline Kernel/Modules.cmake:16]: `if()` interprets a whole host of strings as "truthy" and "falsey" https://cmake.org/cmake/help/latest/command/if.html#basic-expressions

> True if the constant is 1, ON, YES, TRUE, Y, or a non-zero number. False if the constant is 0, OFF, NO, FALSE, N, IGNORE, NOTFOUND, the empty string, or ends in the suffix -NOTFOUND. Named boolean constants are case-insensitive. 

So we could instead `set(COMPILE_FOO ON)` or `set(COMPILE_FOO TRUE)` or some such, and then put any of those strings in the .ini file. even "Y" (not 100% sure if "y" works), and then all the checks below become one of: 

```
if (COMPILE_FOO)
   // compile foo
endif()
```
or 
```
if ("${COMPILE_FOO}")
   // compile foo
endif()
```
but I think the first one should work. 

if "y" doesn't work and you want to keep it lowercase in the .ini file, then a simple `string(TOUPPER "${value}" value)` should do the trick

File: Meta/Azure/Serenity.yml [hunk 6]
```diff
@@ -37,6 +37,7 @@ jobs:
         cmake -S Meta/CMake/Superbuild -B Build/superbuild -GNinja \
           -DSERENITY_ARCH=${{ parameters.arch }} \
           -DSERENITY_TOOLCHAIN=Clang \
+          -DENABLE_ALL_KERNEL_MODULES=ON \
```
Comment (ADKaster) [inline Meta/Azure/Serenity.yml:40]: Is the intention to always boot in CI with all modules? or just make sure they compile?

If we only care about compilation then adding `set(ENABLE_ALL_KERNEL_MODULES ON)` to the top level CMakeLists.txt instead of to each CI job would be cleaner

(line 114 in latest master)
```
if (ENABLE_ALL_DEBUG_FACILITIES)
    set(ENABLE_ALL_THE_DEBUG_MACROS ON)
    set(ENABLE_EXTRA_KERNEL_DEBUG_SYMBOLS ON)
```

**Review by ADKaster — COMMENTED** [review]
I remember you mentioning that you still wanted to look into the `initialize_if()` construct, but here's some CMake comments :D

File: Meta/Azure/Serenity.yml (hunk 6 above)
Comment (supercomputer7) [reply Meta/Azure/Serenity.yml:40]: The initial intention was indeed to ensure the code doesn't rot, but I also want to ensure we don't make it to rot in runtime, hence we try to make the best effort we can to ensure it boots OK with all modules being compiled into the Kernel.

File: Kernel/CMakeLists.txt (hunk 3 above)
Comment (supercomputer7) [reply Kernel/CMakeLists.txt:434]: You're absolutely right about this. I will also look into how to move the file to the Meta folder, although it might a bit of work. We will see how it goes :)

File: Meta/Azure/Serenity.yml (hunk 6 above)
Comment (supercomputer7) [reply Meta/Azure/Serenity.yml:40]: If you feel like I didn't explain it well enough, feel free to unresolve this conversation.

File: Kernel/Modules.cmake (hunk 5 above)
Comment (supercomputer7) [reply Kernel/Modules.cmake:16]: That's quite nice. I'll try to make it to work then :)

File: Kernel/Net/NetworkingManagement.cpp (hunk 2 above)
Comment (supercomputer7) [reply Kernel/Net/NetworkingManagement.cpp:96]: So, I tried to find a way to create a `initalize_if(COMPILE_XXXX)` macro, but without much success. I think I saw the nouveau project puts some macros to take another "macro" and use that like the way we want. Still, it becomes not elegant that much to the point I rather put a few ifdefs in the project (and you can see I put a very small amount of them right now) instead of doing that.
So, either we keep the ifdefs and try to find something better in the future or try the last 2 options. The third option is somewhat better imho than the fourth option which I really dislike (I really don't want to mess around with global constructors). Still, I'm not sure I'm 100% positive on how to implement the third option - what will call the registering function I just compiled in the Kernel?

TL;DR - I need a bit more time with this. I might call it a day sometime soon because I don't want to rebase this pull request forever, and it essentially works (minus all the crappiness I did with build files, pardon my lack of knowledge with that sector of the project). However, I don't think ifdefs are great either, so even in the case where we merge this, I'll definitely put some time into this and eventually I'll find a solution to this.

Comment (supercomputer7) [reply Kernel/Net/NetworkingManagement.cpp:96]: Just to put things in good perspective - this PR adds 27 ifdefs to the whole code base. If you look where I added those ifdefs, they are mainly in initialization code (i.e. in Singletons), and they don't have huge blocks within them, like other ifdefs that are already present in the Kernel code... so still, not a great solution, but I'd not treat them as completely and utterly bad pattern if used wisely (so we don't put ifdef on a large block of code... the largest block of code that can be disabled by an ifdef is a block of 9 lines of code right now).

File: Kernel/Modules.cmake (hunk 5 above)
Comment (supercomputer7) [reply Kernel/Modules.cmake:16]: Before I apply this suggestion I want to ensure I fix the bugs I created with Clang CI so it is marked green before I change anything else. Then I'll fix this, and hopefully after that I'll figure out the ifdefs and this will be ready for being merged :)

Comment (supercomputer7) [reply Kernel/Modules.cmake]: This seems to work out quite nicely, thank you again for the suggestion! :)

**Comment by supercomputer7** [comment]
The last thing to do is to find what to do with the ifdefs. I'll try to give it a couple of days and we will see what happens.
I mark this as being ready to review (i.e. not draft), very happy to see these changes coming to a final phase now :)
There will be more PRs on this subject for sure (to expand the modularization of the Kernel and to fine-tune the build files if needed). I expect to make more kernel drivers to be in the state where you can easily opt-out from compiling them.
This includes the USB subsystem, VMMouse [VMWare Virtual Machine PS2 mouse], interrupt controllers (for example, on arm64 you don't really want to include the IOAPIC nor the PIC, but use whatever available for the board), the SB16 card (and when we have more drivers in the sound subsystem, to include them), the VirtIO subsystem.
More radical ideas include the ACPI and PCI subsystems (so for arm64 boards which don't support these components, we don't need them), but we will see if that's actually needed :)

cc @awesomekling @ADKaster

**Review by friendlyanon — CHANGES_REQUESTED** [review]

File: Kernel/Modules.cmake (hunk 5 above)
Comment (friendlyanon) [reply Kernel/Modules.cmake]: FYI, the boolan constants are checked case insensitively, except for `NOTFOUND`.

File: Kernel/Modules.cmake [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+message("Compiling Intel graphics support was set to " "${COMPILE_INTEL_GRAPHICS}" )
+if (COMPILE_INTEL_GRAPHICS)
+    add_library(INTEL_GRAPHICS OBJECT
+        Graphics/Intel/NativeGraphicsAdapter.cpp
+    )
+    target_compile_definitions(Kernel PRIVATE COMPILE_INTEL_GRAPHICS)
+    target_link_libraries(Kernel PUBLIC INTEL_GRAPHICS)
+endif()
+
+# FIXME: Find a way to use target_compile_definitions instead of add_compile_definitions(COMPILE_VIRTIO_GRAPHICS)
```
[the remark this answers is not in the document]
Comment (supercomputer7) [reply Kernel/Modules.cmake:141]: I made a few changes and resolved this problem now :)

File: Kernel/Net/NetworkingManagement.cpp (hunk 2 above)
Comment (supercomputer7) [reply Kernel/Net/NetworkingManagement.cpp:96]: I'll give it a bit more time, but I reduced the amount of the ifdefs even more now. It's getting harder for rebasing this pull request as more changes are pushed to master.
Marking this as resolved, although I know this is certainly not the case but we can investigate into this in the near future :)

**Comment by supercomputer7** [comment]
I consider this as ready for final review and merging (if seems appropriate).

cc @awesomekling

**Comment by supercomputer7** [comment]
I keep rebasing on & on, and I'll work on it later on.
For now closing this.
